src_features/signMessageEIP712/field_hash.c

```diff
@@ -0,0 +1,284 @@
+#include "ui_logic.h"
+#include "ethUtils.h"        // KECCAK256_HASH_BYTESIZE
+#include "context_712.h"     // contract_addr
+#include "utils.h"           // u64_from_BE
+#include "apdu_constants.h"  // APDU response codes
+#include "typed_data.h"
+#include "commands_712.h"
+#include "hash_bytes.h"
+
+static s_field_hashing *fh = NULL;
+
+/**
+ * Initialize the field hash context
+ *
+ * @return whether the initialization was successful or not
+ */
+bool field_hash_init(void) {
+    if (fh == NULL) {
+        if ((fh = MEM_ALLOC_AND_ALIGN_TYPE(*fh)) == NULL) {
+            apdu_response_code = APDU_RESPONSE_INSUFFICIENT_MEMORY;
+            return false;
+        }
+        fh->state = FHS_IDLE;
+    }
+    return true;
+}
+
+/**
+ * Deinitialize the field hash context
+ */
+void field_hash_deinit(void) {
+    fh = NULL;
+}
+
+/**
+ * Special handling of the first chunk received from a field value
+ *
+ * @param[in] field_ptr pointer to the struct field definition
+ * @param[in] data the field value
+ * @param[in,out] data_length the value length
+ * @return the data pointer
+ */
+static const uint8_t *field_hash_prepare(const void *const field_ptr,
+                                         const uint8_t *data,
+                                         uint8_t *data_length) {
+    e_type field_type;
+
+    field_type = struct_field_type(field_ptr);
+    fh->remaining_size = __builtin_bswap16(*(uint16_t *) &data[0]);  // network byte order
+    data += sizeof(uint16_t);
+    *data_length -= sizeof(uint16_t);
+    fh->state = FHS_WAITING_FOR_MORE;
+    if (IS_DYN(field_type)) {
+        cx_keccak_init(&global_sha3, 256);  // init hash
+        ui_712_new_field(field_ptr, data, *data_length);
+    }
+    return data;
+}
+
+/**
+ * Finalize static field hash
+ *
+ * Encode the field data depending on its type
+ *
+ * @param[in] field_ptr pointer to the struct field definition
+ * @param[in] data the field value
+ * @param[in] data_length the value length
+ * @return pointer to the encoded value
+ */
+static const uint8_t *field_hash_finalize_static(const void *const field_ptr,
+                                                 const uint8_t *const data,
+                                                 uint8_t data_length) {
+    uint8_t *value = NULL;
+    e_type field_type;
+
+    field_type = struct_field_type(field_ptr);
+    switch (field_type) {
+        case TYPE_SOL_INT:
+            value = encode_int(data, data_length, get_struct_field_typesize(field_ptr));
+            break;
+        case TYPE_SOL_UINT:
+            value = encode_uint(data, data_length);
+            break;
+        case TYPE_SOL_BYTES_FIX:
+            value = encode_bytes(data, data_length);
+            break;
+        case TYPE_SOL_ADDRESS:
+            value = encode_address(data, data_length);
+            break;
+        case TYPE_SOL_BOOL:
+            value = encode_boolean((bool *) data, data_length);
+            break;
+        case TYPE_CUSTOM:
```

Comment:
_TYPE_CUSTOM_ means a struct type, which makes no sense in this encoding function since we don't encode a struct, but instead all of its properties that have predefined types.